File: backend/utils/encrypt.js
const crypto = require("crypto");
const dotenv = require("dotenv");
dotenv.config();

class cryptoUtils {
  static SECRET_KEY = Buffer.from(
    process.env.ENC_SECRET_KEY || "12345678901234567890123456789012" // 32 bytes
  );

  static encryptApiKey(apiKey) {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv(
      "aes-256-cbc",
      this.SECRET_KEY,
      iv
    );

    let encrypted = cipher.update(apiKey, "utf8");
    encrypted = Buffer.concat([encrypted, cipher.final()]);

    return Buffer.concat([iv, encrypted]).toString("base64");
  }

  static decryptApiKey(encKey) {
    const raw = Buffer.from(encKey, "base64");
    const iv = raw.slice(0, 16);
    const encrypted = raw.slice(16);

    const decipher = crypto.createDecipheriv(
      "aes-256-cbc",
      this.SECRET_KEY,
      iv
    );

    let decrypted = decipher.update(encrypted);
    decrypted = Buffer.concat([decrypted, decipher.final()]);

    return decrypted.toString("utf8");
  }
}

module.exports = cryptoUtils;

File: fastapi/utils/encrypt.py
# encrypting & decrypting api keys
import base64
from Crypto.Cipher import AES
import os

class CryptoUtils:
    def __init__(self):
        self.SECRET_KEY = os.getenv("ENC_SECRET_KEY", "12345678901234567890123456789012").encode()  # 32 bytes
        self.BLOCK_SIZE = 16

class Encryptor(CryptoUtils):
    def __init__(self):
        super().__init__()
    def pad(self,data: bytes) -> bytes:
        pad_len = self.BLOCK_SIZE - len(data) % self.BLOCK_SIZE
        return data + bytes([pad_len] * pad_len)

    def encrypt_api_key(self,api_key: str) -> str:
        iv = os.urandom(16)
        cipher = AES.new(self.SECRET_KEY, AES.MODE_CBC, iv)
        enc = cipher.encrypt(self.pad(api_key.encode()))
        return base64.b64encode(iv + enc).decode()

class Decryptor(CryptoUtils):
    def __init__(self):
        super().__init__()
    def unpad(self,data: bytes) -> bytes:
        return data[:-data[-1]]

    def decrypt_api_key(self,enc_key: str) -> str:
        raw = base64.b64decode(enc_key)
        iv, enc = raw[:16], raw[16:]
        cipher = AES.new(self.SECRET_KEY, AES.MODE_CBC, iv)
        dec = self.unpad(cipher.decrypt(enc))
        return dec.decode()